File: site_ws/src/site_model/src/data_output/radar_listener.cpp
/*###########################################################
#   This cpp file subscribe the topic of radars and save    #
#	the poses information of vehicles.						#
#   Author: Yangxiuyu                                       #
############################################################*/

#include "ros/ros.h"
#include "std_msgs/String.h"
#include <pcl/point_cloud.h> 
#include <pcl_conversions/pcl_conversions.h> 
#include <sensor_msgs/PointCloud2.h> 
#include <per_msgs/SensorMsgsRadar.h>
#include <per_msgs/GeometryMsgsRadarObject.h>
#include <vector>
#include <fstream>
#include <iostream>
#include <time.h>
#include <string>

using namespace std;

void RadarCB(const per_msgs::SensorMsgsRadar &Radarmsg)
{
    time_t currentTime=time(NULL);
	char chCurrentTime[256];
    

	std::cerr << "Successfully" << std::endl;
	ros::NodeHandle nh; 
	ros::Publisher radar_pub = nh.advertise<sensor_msgs::PointCloud2> ("/radar_show_2", 1);     
	pcl::PointCloud<pcl::PointXYZ> cloud; 
	sensor_msgs::PointCloud2 output; 
	
	// Fill in the cloud data 
	int v_num = Radarmsg.front_left_esr_tracklist.size();
	std::cerr << "left_tracklist_size: " << Radarmsg.front_left_esr_tracklist.size() << std::endl;
	if (v_num!=0)
	{
		// Set the file name
		strftime(chCurrentTime,sizeof(chCurrentTime),"%Y%m%d %H%M%S",localtime(&currentTime));
		string stCurrentTime=chCurrentTime;
		string filename="/home/zonlin/IPP_WorkSpace/ROS_WS/site_ws/src/site_model/dataset/radar_data/data"+stCurrentTime+".txt";

		cloud.width = v_num; 
		cloud.height = 1; 
		cloud.points.resize(cloud.width * cloud.height); 

		for(int i=0;i<v_num;i++)
		{
			cloud.points[i].x = Radarmsg.front_left_esr_tracklist[i].obj_vcs_posex;
			cloud.points[i].y = Radarmsg.front_left_esr_tracklist[i].obj_vcs_posey;
			cloud.points[i].z = 2; 
		}

		// Convert the cloud to ROS message 
		pcl::toROSMsg(cloud, output); 
		output.header.frame_id = "/base_link"; 

		// Publish the data
		radar_pub.publish(output);

		// Save data to txt files
		ofstream fout;
		fout.open(filename.c_str());
		int i=0;
		while(i<v_num)
		{
			fout<< i << " " << cloud.points[i].x <<" "<< cloud.points[i].y << endl;
			i++;
		}
		fout<<flush;
		fout.close();
	}
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "radar_listener");
    ros::NodeHandle nh;
    ros::Subscriber sub = nh.subscribe("/ARS_408_21_2_Topic", 10, RadarCB);
    ros::spin();
    return 0;
}

File: site_ws/src/site_model/src/tools/radar_listener.py
#############################################################
#   This py file subscribe the topic of radars and save     #
#	the poses information of vehicles.						#
#   Author: Yangxiuyu                                       #
#############################################################

from sqlalchemy import true
import rospy
from per_msgs.msg._SensorMsgsRadar import *
# radar message type
from msgs.msg._MsgRadar import *
from msgs.msg._MsgRadarObject import *
import message_filters

def radar_listener(radar2: SensorMsgsRadar, radar3: SensorMsgsRadar):
    msgradar = MsgRadar()
    if radar2.total_front_left_esr_tracks != 0: # if num of vehicles is not 0
        msgradar.total_vehicles_left = round(radar2.total_front_left_esr_tracks)
        for i in range(msgradar.total_vehicles_left):
           object = MsgRadarObject() # define a new list
           object.obj_vcs_posex = radar2.front_left_esr_tracklist[i].obj_vcs_posex
           object.obj_vcs_posey = radar2.front_left_esr_tracklist[i].obj_vcs_posey
           object.range_rate = radar2.front_left_esr_tracklist[i].range_rate
           msgradar.ObjectList_left.append(object)

    if radar3.total_front_right_esr_tracks !=0:
        msgradar.total_vehicles_right = round(radar3.total_front_right_esr_tracks)
        for i in range(msgradar.total_vehicles_right):
           object = MsgRadarObject()
           object.obj_vcs_posex = radar3.front_right_esr_tracklist[i].obj_vcs_posex
           object.obj_vcs_posey = radar3.front_right_esr_tracklist[i].obj_vcs_posey
           object.range_rate = radar3.front_right_esr_tracklist[i].range_rate
           msgradar.ObjectList_right.append(object)

    pub = rospy.Publisher("/radar_msgs_combined", MsgRadar) # MsgRadar is the defined message type
    msgradar.header.stamp = rospy.Time.now() # Add time stamp
    pub.publish(msgradar)

if __name__ == '__main__':
    rospy.init_node("radar_listener", anonymous=true)

    sub_radar_2 = message_filters.Subscriber("/ARS_408_21_2_Topic", SensorMsgsRadar)
    sub_radar_3 = message_filters.Subscriber("/ARS_408_21_3_Topic", SensorMsgsRadar)

    sync = message_filters.ApproximateTimeSynchronizer([sub_radar_2, sub_radar_3], 1, 1)

    sync.registerCallback(radar_listener)
    print("Radar Listener Begin.")
    rospy.spin()